src/server/pkg/worker/branch_set.go

```diff
@@ -0,0 +1,245 @@
+
+	commitSets := make([][]*pfs.CommitInfo, len(directInputs))
+	var commitSetsMutex sync.Mutex
+
+	ch := make(chan *branchSet)
+	for i, input := range directInputs {
+		i, input := i, input
+
+		request := &pfs.SubscribeCommitRequest{
+			Repo:   client.NewRepo(input.Repo),
+			Branch: input.Branch,
+		}
+		if input.FromCommit != "" {
+			request.From = client.NewCommit(input.Repo, input.FromCommit)
+		}
+		stream, err := pfsClient.SubscribeCommit(ctx, request)
+		if err != nil {
+			cancel()
+			return nil, err
+		}
+
+		go func() {
+			for {
+				commitInfo, err := stream.Recv()
+				if err != nil {
+					select {
+					case <-ctx.Done():
+					case ch <- &branchSet{
+						Err: err,
+					}:
+					}
+					return
+				}
+
+				commitSetsMutex.Lock()
+				commitSets[i] = append(commitSets[i], commitInfo)
+
+				// Now we look for a set of commits that have provenance
+				// commits from the root input repos.  The set is only valid
+				// if there's precisely one provenance commit from each root
+				// input repo.
+				if set := considerEachSet(commitSets, i, func(set []*pfs.CommitInfo) bool {
+					rootCommits := make(map[string]map[string]bool)
+					setRootCommit := func(commit *pfs.Commit) {
+						if rootCommits[commit.Repo.Name] == nil {
+							rootCommits[commit.Repo.Name] = make(map[string]bool)
+						}
+						rootCommits[commit.Repo.Name][commit.ID] = true
+					}
+					for _, commitInfo := range set {
+						setRootCommit(commitInfo.Commit)
+						for _, provCommit := range commitInfo.Provenance {
+							setRootCommit(provCommit)
+						}
+					}
+					// ok tells us if it's ok to spawn a job with this
+					// commit set.
+					for _, rootInput := range rootInputs {
+						if len(rootCommits[rootInput.Repo]) != 1 {
+							return false
+						}
+					}
+					return true
+				}); set != nil {
+					bs := &branchSet{
+						NewBranch: i,
+					}
+					for _, commitInfo := range set {
+						bs.Branches = append(bs.Branches, &pfs.Branch{
+							Name: input.Branch,
+							Head: commitInfo.Commit,
+						})
+					}
+					select {
+					case <-ctx.Done():
+						commitSetsMutex.Unlock()
+						return
+					case ch <- bs:
+					}
+				}
+				commitSetsMutex.Unlock()
+			}
+		}()
+	}
+
+	f := &branchSetFactoryImpl{
+		cancel: cancel,
+		ch:     ch,
+	}
+
+	return f, nil
+}
+
+// considerEachSet runs a function on every commit set, starting with the
+// most recent one.  If the function returns true, then considerEachSet
+// removes the commit sets that are older than the current one and returns
+// the current one.
+// i is the index of the input from which we just got a new commit.  Since
+// it's the "triggering" input, we know that we only have to consider commit
+// sets that include the triggering commit since other commit sets must have
+// already been considered in previous runs.
+func considerEachSet(commitSets [][]*pfs.CommitInfo, i int, f func(commitSet []*pfs.CommitInfo) bool) []*pfs.CommitInfo {
```

Comment:
Could we name this something to indicate that it's doing provenance matching? `considerEachSet` seems a bit too generic to me because it doesn't tell me what it's actually considering about them.